### PR TITLE
Fix bad indent on align double arrow fixer

### PR DESCRIPTION
**Pull request by Soullivaneuh**
This is the same as #1463, I did something wrong with git.

### CODE REVIEW

**Comment by Soullivaneuh** [comment]
@keradus tests passed, ready to merge. :+1:

**Comment by keradus** [comment]
I would like to get answer for https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1463#discussion_r43188994 first
@cc GrahamCampbell

**Comment by Soullivaneuh** [comment]
> I would like to get answer for #1463 (comment) first

I can understand, but this annoying bug is here from a long time and the fix is ready.

New tests passed and no regression with other tests, what is missing?

I would like to add tests later if @GrahamCampbell want it but I think this PR can be merged as is.

File: Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php [hunk 1]
```diff
@@ -139,7 +139,7 @@ private function injectAlignmentPlaceholders(Tokens $tokens, $startAt, $endAt)
                 continue;
             }
 
-            if ($token->equals(',')) {
+            if ($token->equals(',') && !in_array($tokens[$index - 1]->getContent(), array(')', ']'), true)) {
```
Comment (keradus) [inline Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php]: use `equalsAny`

**Comment by keradus** [comment]
> New tests passed and no regression with other tests, what is missing?

Votes with approval for this PR. Fix probably is good, but most of us just need more time than few hours to think about the bug, solution and discover edge cases

File: Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php (hunk 1 above)
Comment (Soullivaneuh) [inline Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php]: Don't know about this method, will fix.

BTW, phpdoc should be improved to have better completion.

e.g.: `Tokens|Token[]`.

**Comment by Soullivaneuh** [comment]
@keradus review corrected.

File: Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php (hunk 1 above)
Comment (keradus) [inline Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php]: `Tokens` or `Token[]` or array of token prototype, where each prototype may be as well array like string.
eg:
```php
array(
    new Token(...),
    array(T_STRING, 'foo'),
    '('
)
```

Comment (Soullivaneuh) [inline Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php]: Well, `Tokens|Token[]|array` so. :-)

Comment (keradus) [inline Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php]: :+1:

**Comment by Soullivaneuh** [comment]
@SpacePossum Could you please take a look of this one? Thanks!

**Comment by GrahamCampbell** [comment]
What I was saying is you need to check these tokens actually belong to an array rather than:

`$foo[1],` where that would match a closing bracket and a comma.

**Comment by GrahamCampbell** [comment]
Looks like we're missing a test case:

```php
<?php

function foo() {}

$bar = 42;

$foo = [
    'foo' => $bar[123],
    'a' => foo(),
];
```

**Comment by GrahamCampbell** [comment]
If that gets fixed correctly after your change, than that's fine. I'd just like to be sure. :)

**Comment by Soullivaneuh** [comment]
@GrahamCampbell I added your test, seems to be OK. :+1: 

cc @keradus

**Comment by keradus** [comment]
looking on the condition, could you change the case into:

```diff
    $foo = [
        "test123" => "foo",
        "foo" => $bar[123],
        "a" => foo(),
+       "b" => 1,
    ];
```

?

**Comment by GrahamCampbell** [comment]
:+1:

**Comment by Soullivaneuh** [comment]
@keradus done.

**Comment by SpacePossum** [comment]
Thanks for the PR. I've been testing a bit with it and I'm wondering what people think of the following two failing test cases.
Kindly note, these fail with and without your changes.

```php
            array(
                '<?php
    return array(
        self::STATUS_UNKNOWN    => array("symbol" => "?", "description" => "unknown"),
        self::STATUS_INVALID    => array("symbol123" => "III", "description" => "invalid file syntax, file ignored"),
    );',
                '<?php
    return array(
        self::STATUS_UNKNOWN => array("symbol" => "?", "description" => "unknown"),
        self::STATUS_INVALID    => array("symbol123" => "III", "description" => "invalid file syntax, file ignored"),
    );',
```

```php
            array(
                '<?php
    return array(
        self::STATUS_UNKNOWN    => array((1+11)=> "?", "description" => "unknown"),
        self::STATUS_INVALID    => array((2+3)=> "III", "description" => "invalid file syntax, file ignored"),
    );',
                '<?php
    return array(
        self::STATUS_UNKNOWN => array((1+11)=> "?", "description" => "unknown"),
        self::STATUS_INVALID    => array((2+3)=> "III", "description" => "invalid file syntax, file ignored"),
    );',
            ),
```

In other words, the inline array detection does not seem to work for all cases I think.

**Comment by Soullivaneuh** [comment]
Thanks @SpacePossum, will take a look.

**Comment by Soullivaneuh** [comment]
Sorry @SpacePossum, but what is the difference between your two pastes?

**Comment by SpacePossum** [comment]
in case 1, hard index, space between '3"' and '=>' :
`array("symbol123" => "III",`

in case 2, 'dynamic' index, no space between ')' and '=>' :
`array((2+3)=> "III",`

**Comment by Soullivaneuh** [comment]
Oh, didn't saw it. Need a :coffee:. :-)

**Comment by Soullivaneuh** [comment]
I'm trying to check with line position:

```php
            if ($token->isGivenKind(T_ARRAY)) { // don't use "$tokens->isArray()" here, short arrays are handled in the next case
                $from = $tokens->getNextMeaningfulToken($index);
                $lineFrom = $tokens[$from]->getLine();
                $until = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_PARENTHESIS_BRACE, $from);
                $lineUntil = $tokens[$until]->getLine();
                $index = $until;

                var_dump($lineFrom, $lineUntil);
                if ($lineUntil > $lineFrom) {
                    ++$this->deepestLevel;
                    ++$this->currentLevel;
                    $this->injectAlignmentPlaceholders($tokens, $from, $until);
                    --$this->currentLevel;
                }

                continue;
            }
```

But `getLine` always returns `NULL`.

@keradus do you know why?

**Comment by keradus** [comment]
Yes. Never, never use `getLine`. As token prototype from `tokens_get_all` has it, `Token` class has it too. But when all our fixers do theirs stuff they change lines, adding new, removing old, and _do not_ update all lines of all tokens in collection. Do not rely on that method.
After said that, the method was removed on 2.x line.

**Comment by Soullivaneuh** [comment]
Well, that's so bad. How can I deal with line? I don't know how to solve it with other way.

PHP code sniffer tokenizer have line management working well, maybe could we get inspiration of it?

**Comment by keradus** [comment]
So far we need to handle line numbers in so few places, while keeping them up to date was costly, we decided to throw it away.
In that few places if we need to know if the tokens are in same line or not - we simply analyze content between, eg https://github.com/Soullivaneuh/PHP-CS-Fixer/blob/issue-1450-proof/Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php#L158

**Comment by Soullivaneuh** [comment]
@SpacePossum @keradus fixed! :+1:

File: Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php [hunk 2]
```diff
@@ -112,10 +116,14 @@ private function injectAlignmentPlaceholders(Tokens $tokens, $startAt, $endAt)
                 $until = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_SQUARE_BRACE, $from);
                 $index = $until;
 
-                ++$this->deepestLevel;
-                ++$this->currentLevel;
-                $this->injectAlignmentPlaceholders($tokens, $from + 1, $until - 1);
-                --$this->currentLevel;
+                // Inject placeholders only for multi lines array
+                if (false !== strpos($tokens->generatePartialCode($from, $until), "\n")) {
```
Comment (GrahamCampbell) [inline Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php]: maybe this needs abstracting to a method?

**Comment by Soullivaneuh** [comment]
@GrahamCampbell suggestion added.

File: Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php (hunk 2 above)
Comment (keradus) [inline Symfony/CS/Fixer/Contrib/AlignDoubleArrowFixer.php]: https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1466#issuecomment-152152097 ;P

**Comment by Soullivaneuh** [comment]
@keradus Are we now? :-)

**Comment by keradus** [comment]
LGTM, just pleae get rid of that merge commit (eg rebase on 1.10)

**Comment by Soullivaneuh** [comment]
Indeed, nicer history. :-)

PR ready to merge! :+1: 

I'm curious: Why a simple `rebase 1.10` can get rid of merge commits? And what would happen if I was already up to date from `1.10`?

**Comment by keradus** [comment]
After rebasing on 1.10 the merge commit wouldn't bring any code changes (because you are working on top of 1.10), therefore it will be considered as empty commit, and while rebasing empty commits are ommited by default.
If your branch would be created on top of 1.10 - then you wouldn't have merge commit of 1.10 (that would mean the 1.10 has new commits and your branch is not created on top of 1.10).

**Comment by keradus** [comment]
:+1:

**Comment by SpacePossum** [comment]
looking good :+1:

**Comment by Soullivaneuh** [comment]
Thanks. :-)

When this could be merged?

**Comment by keradus** [comment]
Thank you @Soullivaneuh.

**Comment by SpacePossum** [comment]
thanks @Soullivaneuh !

**Comment by Soullivaneuh** [comment]
Thank you guys! :+1:

**Comment by Soullivaneuh** [comment]
@keradus All PR seems to be [merged](https://github.com/FriendsOfPHP/PHP-CS-Fixer/milestones/v1.10.3), do you have a date for v1.10.3 release?

**Comment by keradus** [comment]
I was already trying to do it for few days... I have just change the country, please don't be impatient ;)
